README
NAME
    Net::Telnet::Trango - Perl extension for accessing the Trango telnet
    interface

SYNOPSIS
      use Net::Telnet::Trango;
      my $t = new Net::Telnet::Trango ( Timeout => 5 );

      $t->open( Host => $ap ) or die "Error connecting: $!";

      $t->login('password') or die "Couldn't log in: $!";

      # Do whatever

      $t->exit;
      $t->close;

DESCRIPTION
    Perl access to the telnet interface on Trango APs and SUs.

    A handy feature is that it will parse the output from certain commands
    that is in the format "[key1] value1 [key2] value2" and put those in a
    hashref that is returned. This makes using the output from things like
    sysinfo very easy to do.

  EXPORT
    None

METHODS
  new - Creates a new Net::Telnet::Trango object.
        new([Options from Net::Telnet,]
            [Decode => 0,]);

    Same as new from Net::Telnet but sets the default Trango Prompt:
    '/[\$#]>\s*\Z/'

    It also takes an optional parameter 'Decode'. If not defined it defaults
    to 1, if it is set to 0, it will not decode the output and instead
    return a reference to an array of the lines that were returned from the
    command.

ACCESSORS
    These are usually only set internally.

  firmware_version - returns the firmware version
    Returns the firmware version if available, otherwise undef.

    It should be available after a successful open().

  host_type - return the type of host you are connected to.
    returns the type of host from the login banner for example M5830S or
    M5300S.

    Should be available after a successful open().

  is_connected - Status of the connection to host.
    returns 1 when connected, undef otherwise.

  logged_in - Status of being logged in to the host.
    returns 1 after a successful login(), 0 if it failed and undef if
    login() was never called.

  login_banner - The banner when first connecting to the host.
    returns the banner that is displayed when first connected at login. Only
    set after a successful open().

  last_lines - The last lines of output from the last cmd().
    returns, as an array ref, the output from the last cmd() that was run.

  last_error - A text output of the last error that was encountered.
    returns the last error reported. Probably contains the last entry in
    last_lines.

ALIASES
  bye - alias of exit()
    Does the same as exit()

  restart - alias of reboot()
    Does the same as reboot()

  save_systemsetting - alias of save_ss()
    Does the same as save_ss()

COMMANDS
    Most of these are just shortcuts to "cmd(String => METHOD)", as such
    they accept the same options as "cmd()". Specifically they take a named
    paramater "args", for example: "tftpd(args => 'on')" would enable tftpd

  tftpd - The output from the tftpd command
    Returns a hash ref of the decoded output from the command.

    Also see enable_tftpd() and disable_tftpd() as those check that it was
    successfully changed.

  ver - The output from the ver command
    Returns a hash ref of the decoded output from the command.

  sysinfo - The output from the sysinfo command
    Returns a hash ref of the decoded output from the command.

  exit - Exits the connection
    exits the command session with the Trango and closes the connection

  reboot - Sends a reboot command
    reboots the Trango and closes the connection

  remarks - Set or retrieve the remarks.
    Takes an optional argument, which sets the remarks. If there is no
    argument, returns the current remarks.

      my $old_remarks = $t->remarks();
      $t->remarks($new_remarks);

  sulog - The output from the sulog command
    Returns an array ref of hashes containing each log line.

  save_sudb - saves the sudb
    Returns true on success, undef on failure

  syslog - The output from the sulog command
    Returns a hashref of the output from the syslog command

  pipe - the pipe command
    Returns the output from the pipe command

  maclist - retrieves the maclist
    Returns the output from the maclist command

  maclist_reset - resets the maclist.
    No useful output.

  eth_link - eth link command
    Returns the output from the eth link command

    This command seems to cause some weird issues. It often will cause the
    command after it to appear to fail. I am not sure why.

  su_info - gets the su info
    Returns information about the SU.

    You need to pass in the $suid and it will return the info for that suid.

      $t->su_info($suid);

  su_testrflink - tests the RF Link to an su
      $t->su_testrflink($suid|'all');

  save_ss - saves the config.
    Returns 1 on success, undef on failure.

  opmode - sets opmode ap y or returns the opmode
        $t->opmode([ap y]);

  open - Open a connection to a Trango AP.
    Calls Net::Telnet::open() then makes sure you get a password prompt so
    you are ready to login() and parses the login banner so you can get
    host_type() and firmware_version()

  login - Login to the AP.
    Calls open() if not already connected, then sends the password and sets
    logged_in() if successful

  parse_login_banner - Converts the login_banner to something useful.
    Takes a login banner (what you get when you first connect to the Trango)
    or reads what is already in login_banner() then parses it and sets
    host_type() and firmware_version() as well as login_banner()

  linktest - Link test to SU
    linktest('suid'[, 'pkt len, bytes'[, '# of pkts'[, '# of cycles']]]);

    Returns a hash reference to the results of the test

  su_password - Set the password on SUs connected to the AP.
    su_password('new_password'[, 'suid']) If no suid is specified, the
    default is "all".

      $t->su_password('good_pass', 5);

  su_ipconfig - Change IP configuration on SUs connected to the AP.
    su_ipconfig( 'suid', 'new_ip', 'new_subnet', 'new_gateway' )

      $t->su_ipconfig( 5, '10.0.1.5', '255.255.255.0', '10.0.1.1' );

  sudb_view - Returns the output from the sudb view command
    returns a reference to an array of hashes each containing these keys
    'suid', 'su2su', 'type', 'cir', 'mir' and 'mac'

  sudb_add - Adds an su to the sudb
    Takes the following paramaters

        suid : numeric,
        type : (reg|pr)
        cir  : numeric,
        mir  : numeric,
        mac  : Almost any format, it will be reformatted,

    and returns true on success or undef otherwise.

      $t->sudb_add($suid, 'reg', $cir, $mir, $mac);

    You should save_sudb() after calling this, or your changes will be lost
    when the AP is rebooted.

  sudb_delete - removes an su from the sudb
    Takes either 'all' or the suid of the su to delete and returns true on
    success or undef otherwise.

      $t->sudb_delete($suid);

    You should save_sudb() after calling this, or your changes will be lost
    when the AP is rebooted.

  sudb_modify - changes the su information in the sudb
    Takes either the suid of the su to change as well as what you are
    changing, either "cir, mir or su2su" and returns true on success or
    undef otherwise.

    cir and mir also take a value to set the cir/mir to.

    su2su takes a group id parameter that is in hex.

      $t->sudb_modify($suid, 'cir', 512);

    You should save_sudb() after calling this, or your changes will be lost
    when the AP is rebooted.

  enable_tftpd - enable the TFTP server
    runs "tftpd(args => 'on')" and makes sure that Tftpd is now 'listen'ing

  disable_tftpd - disable the TFTP server
    runs "tftpd(args => 'off')" and makes sure that Tftpd is now 'disabled'

  cmd - runs a command on the AP.
    This does most of the work. At the heart, it calls Net::Telnet::cmd()
    but it also does some special stuff for Trango.

    Normally returns the last lines from from the command

    If you are using this, rather than one of the "easy" methods above, you
    probably want to read through the source of this module to see how some
    of the other commands are called.

    In addition to the Net::Telnet::cmd() options, it also accepts these:

    *decode* - if this is true, then it will send the output lines to
    _decode_lines() and then returns the decoded output

    *no_prompt* - if this is true, it does not wait for a prompt, so you are
    not stuck waiting for something that will never happen.

    *cmd_disconnects* - if this is true, it then sets logged_in() to false,
    then it will close() the connection and set is_connected() to false

    *expect* - if this is set (usually to 'Success.') it will check for that
    in the last line of output and if it does not, will return undef because
    the command probably failed

    *args* - a string containing the command line options that are passed to
    the command

        $t->cmd( String => 'exit', no_prompt => 1, cmd_disconnects => 1 );

SEE ALSO
    Trango Documentation -
    <http://www.trangobroadband.com/support/product_docs.htm>

    Net::Telnet

TODO
    There are still a lot of commands that are not accessed directly. If you
    call them (as cmd("command + args") or whatever) and it works, please
    send me examples that work and I will try to get it incorporated into
    the next version of the script.

    I also want to be able to parse the different types of output from
    commands like su, sudb all and anything else that would be better
    available as a perl datastructure.

AUTHOR
    Andrew Fresh <andrew@rraz.net>

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Net::Telnet::Trango

    You can also look for information at:

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/Net-Telnet-Trango>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/Net-Telnet-Trango>

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Net-Telnet-Trango>

    *   Search CPAN

        <http://search.cpan.org/dist/Net-Telnet-Trango>

COPYRIGHT AND LICENSE
    Copyright (C) 2005,2006,2007 by Andrew Fresh

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Net-Telnet-Trango

0.04    2009.07.13
        * Fix a mistake I made in the match for sudb_view. I should probably
          get a test environment set up.  Hopefully that is the last problem
          Steven has to work around for a while.

0.03    2009.07.09
        * Match the su2su field in sudb_view, it is in hex, so \d didn't match
          sometimes.  Thanks again to "Steven N. Severinghaus"
        * More tests that are probably non-functional because I didn't set up
          a test environment.

0.02    2009.07.08
        * Perltidy and some Perl Critic
        * Changed the prompt to include $> not just #>
        * Changed the banner match to support more types of hardware
        * Lots of tests
        * Added stuff to the TODO
        * Support for the TrangoLink 45
        * Additional features in the su.cgi
        * Add commands arq and linktest
        * Fix a problem in sudb_add where $new_mac wasn't completely
          stripped of special characters.
          From: "Steven N. Severinghaus" <steve@volo.net> 

0.01    2007.02.05
        * First version, released on an unsuspecting world.

$RedRiver: Changes,v 1.6 2009/07/13 16:33:45 andrew Exp $


scripts/update_trango.pl
#!/usr/bin/perl
# $RedRiver: update_trango.pl,v 1.32 2007/02/07 19:25:05 andrew Exp $
########################################################################
# update_trango.pl *** Updates trango hosts with a new firmware
# 
# 2005.11.15 #*#*# andrew fresh <andrew@mad-techies.org>
########################################################################
# Copyright (C) 2005, 2006, 2007 by Andrew Fresh
#
# This program is free software; you can redistribute it and/or modify
# it under the same terms as Perl itself.
########################################################################
use strict;
use warnings;

use YAML qw/ LoadFile /;
use Net::TFTP;
use Net::Telnet::Trango;

my $config_file = shift || 'update_trango.yaml';
my $max_tries = 3;

my $l = Mylogger->new( { log_prefix => 'UT' } );

$l->sp("Reading config file '$config_file'");
my $conf = LoadFile($config_file);

my $hosts;
if (@ARGV) {
    @{ $hosts } = map { { name => $_, group => 'Trango-Client' } } @ARGV
} else {
    $hosts = parse_hosts($conf->{hosts});
}

#@{ $hosts } = grep { $_->{name} eq '10.100.7.2' } @{ $hosts };

my $global_tries = $max_tries * 2;
while ($global_tries > 0) { 
    $global_tries--;
    my $processed = 0;

    foreach my $host (@{ $hosts }) {

        if (! exists $host->{retry}) {
            $host->{tries} = 0;
            $host->{retry} = 1;
        }

        if ($host->{tries} >= $max_tries) {
            $host->{retry} = 0;
        }

        if ($host->{retry} <= 0) {
            next;
        }

        $host->{tries}++;
        $processed++;

        $l->sp("");
        $l->sp("Checking: $host->{name} (try $host->{tries})");
        my $needs_reboot = 0;

        ## Connect and login.
        my $t = new Net::Telnet::Trango (
            Timeout => 5,
            Errmode => 'return',
        ) or die "Couldn't make new connection: $!";
        $l->p("Connecting to $host->{name}");
        unless ( $t->open($host->{name}) ) {
            $l->sp("Error connecting: $!");
            next;
        }

        my $password = $host->{Telnet_Password} || $conf->{general}->{password};

        $l->p("Logging in");
        $t->login($password);
        unless ($t->logged_in) {
            $l->p('Failed!');
            $t->close;
            next;
        }

        $l->sp("Getting sudb");
        my $sudb = $t->sudb_view;
        if ($sudb) {
            foreach my $su (@{ $sudb }) {
                $l->p("Getting su info $su->{suid}");
                my $su_info = $t->su_info( $su->{suid} ); 
                if ($su_info->{ip}) {
                    if (grep { $_->{name} eq $su_info->{'ip'} } @{ $hosts }) {
                        $l->p("Already have $su_info->{ip}");
                        next;
                    }
                    $l->sp("Adding host $su_info->{ip}");
                    my $new_host = {
                        password => $host->{password},
                        name     => $su_info->{ip},
                        remarks  => $su_info->{remarks},
                    };
                    push @{ $hosts }, $new_host;
                } else {
                    $l->sp("Couldn't get su info for $su->{suid}");
                    $l->sp("ERR: " . $t->last_error);
                }
            }
        }

        foreach my $firmware_type ('Firmware', 'FPGA') {

            if (! exists $conf->{$firmware_type}) {
                $l->s("No configs for '$firmware_type'");
                next;
            }

            my $host_type = $t->host_type;
            if ($firmware_type eq 'FPGA') {
                $host_type =~ s/\s.*$//;
            }

            if (! exists $conf->{$firmware_type}->{$host_type}) {
                $l->sp("No '$firmware_type' config for type $host_type");
                next;
            }

            if ($firmware_type eq 'Firmware' && 
                $t->firmware_version eq 
                $conf->{$firmware_type}->{$host_type}->{ver}
            ) {
                $l->sp("Firmware already up to date");
                next;
            }

            if (! $t->logged_in) {
                $l->p("Logging in");
                $t->login($password);
                unless ($t->logged_in) {
                    $l->p('Failed!');
                    $t->close;
                    last;
                }
            }

            foreach my $k (keys %{ $conf->{general} }) {
                $conf->{$firmware_type}->{$host_type}->{$k} 
                  ||= $conf->{general}->{$k};
            }
            $conf->{$firmware_type}->{$host_type}->{firmware_type} 
              ||= $firmware_type;
            $conf->{$firmware_type}->{$host_type}->{type} = $host_type;

            $l->sp("$host_type $firmware_type");
            ## Send commands
            my $rc = upload($t, $conf->{$firmware_type}->{$host_type});
            if ($rc) {
                $l->sp("Successfull!");
                $host->{retry}--;
                $needs_reboot = 1;
            } elsif (defined $rc) {
                $l->sp("Already up to date");
                $host->{retry}--;
            } else {
                $l->sp("Failed");
                $t->bye;
                next;
            }

        }

        if ($needs_reboot) {
            $l->sp("Rebooting $host->{name}");
            $t->reboot;
        } else {
            $l->sp("Bye $host->{name}");
            $t->bye();
        }
    }

    if (! $processed) {
        $l->sp("");
        $l->sp("Finished.  No more hosts.");
        last;
    }
}

sub upload
{
    my $t    = shift;
    my $conf = shift;

    my $file = $conf->{firmware_path} . '/' . $conf->{file_name};

    my $fw_type = $conf->{firmware_type};

    my $ver = $t->ver;

    if (! (
            $ver->{$fw_type . ' Version'}  &&
            $ver->{$fw_type . ' Checksum'} 
        )) {
        $l->sp("Error getting current version numbers");
        return;
    }

    if (
        $ver->{$fw_type . ' Version'}  eq $conf->{'ver'} && 
        $ver->{$fw_type . ' Checksum'} eq $conf->{'cksum'}
    ) {
        return 0;
    }

    $l->sp("Updating $fw_type");
    $l->sp("Config information:");
    $l->sp("  Hardware Type: $conf->{'type'}");
    $l->sp("  File Name:     $conf->{'file_name'}");
    $l->sp("  File Size:     $conf->{'file_size'}");
    $l->sp("  File Checksum: $conf->{'file_cksum'}");
    $l->sp("  Conf Version:  $conf->{'ver'}");
    $l->sp("  Cur  Version:  $ver->{$fw_type . ' Version'}");
    $l->sp("  Conf Checksum: $conf->{'cksum'}");
    $l->sp("  Cur  Checksum: $ver->{$fw_type . ' Checksum'}");

    my $try = 0;
    while (1) {
        if ($try >= $max_tries) {
            $l->sp("Couldn't update in $max_tries tries!");
            return;
        }
        $try++;

        $l->p("Enabling TFTPd");
        unless ($t->enable_tftpd) {
            $l->sp("Couldn't enable tftpd");
            next;
        }

        $l->p("Uploading file ($conf->{file_name})");
        # use tftp to push the file up
        my $tftp = Net::TFTP->new($t->Host, Mode => 'octet');

        unless ($tftp->put($file, $file)) {
            $l->sp("Error uploading: " . $tftp->error);
            next;
        }

        $l->p("Checking upload ($conf->{'file_cksum'})");
        my $results = $t->tftpd;
        # check the 'File Length' against ???
        if (! (
                $results->{'File Checksum'} &&
                $results->{'File Length'}   &&
                $results->{'File Name'}
            )) {
            $l->sp("Unable to get results of upload");
            next;
        }
        if ( $results->{'File Checksum'} ne $conf->{'file_cksum'}) {
            $l->sp(
                "File checksum (" . $results->{'File Checksum'} .
                ") does not match config file (" . $conf->{'file_cksum'} . ")!"
            );
            next;
        } 
        $l->p("File checksum matches . . . ");

        if ($results->{'File Length'}   !~ /^$conf->{'file_size'} bytes/) {
            $l->sp(
                "File length (" . $results->{'File Length'} .
                ") does not match config file (" . $conf->{'file_size'} . " bytes)!"
            );
            next;
        }
        $l->p("File length matches . . . ");

        if ( uc($results->{'File Name'}) ne uc($file) ) {
            $l->sp(
                "File name (" . $results->{'File Name'} .
                ") does not match config file (" . $file . ")!"
            );
            next;
        }
        $l->p("File name matches . . . ");

        my $image_type = 'mainimage';
        if ($fw_type eq 'FPGA') {
            $image_type = 'fpgaimage';
        }
        $l->p("Updating $image_type (new checksum '$conf->{'cksum'}')");
        unless ($results = $t->updateflash(
                args => $image_type . ' ' . $ver->{$fw_type . ' Checksum'} . 
                ' '          . $conf->{'cksum'},
                Timeout => 90,
            ) ) {
            $l->sp("Couldn't update flash: $!");
            next;
        }

        unless (
            defined $results->{'Checksum'} && 
            $results->{'Checksum'} eq $conf->{'cksum'}
        ) {
            $l->sp("Saved checksum " . $results->{'Checksum'} . " does not match config file " .  $conf->{'cksum'} . "!");
            next;
        }

        $l->p("$fw_type saved checksum matches . . . ");

        return 1;
    }
}

sub parse_hosts
{
    my $src = shift;

    my @hosts;
    foreach my $h (@{ $src }) {
        if ($h->{name} =~ /^(\d{1,3}\.\d{1,3}\.\d{1,3}\.)(\d{1,3})-(\d{1,3})/) {
            for ($2..$3) {
                my %cur_host;
                foreach my $k (keys %{ $h }) {
                    $cur_host{$k} = $h->{$k};
                }
                $cur_host{name} = $1 . $_;
                if (! grep { $cur_host{name} eq $h->{name} } @hosts) {
                    push @hosts, \%cur_host;
                }
            }
        } else {
            push @hosts, $h;
        }
    }

    return \@hosts;
}

package Mylogger;

use Fcntl ':flock'; # import LOCK_* constants
#use YAML;
use constant LOG_PRINT => 128;
use constant LOG_SAVE  =>  64;

DESTROY {
    my $self = shift;
    if ($self->{'MYLOG'}) {
        $self->p("Closing log ($self->{'log_path'}/$self->{'log_file'})");
        close $self->{'MYLOG'};
    }
}

sub new {
    my $package = shift;
    my $self = shift || {};

    $self->{'base_path'}  ||= '.';
    $self->{'log_path'}   ||= $self->{'base_path'};
    $self->{'log_prefix'} ||= 'LOG';
    $self->{'log_file'}   ||= GetLogName(
        $self->{'log_prefix'}, 
        $self->{'log_path'}
    );
    bless $self, $package;
}

sub s
{
    my $self = shift;
    my $m = shift;
    return $self->mylog($m, LOG_SAVE);
}

sub p
{
    my $self = shift;
    my $m = shift;
    return $self->mylog($m, LOG_PRINT);
}

sub sp
{
    my $self = shift;
    my $m = shift;
    return $self->mylog($m, LOG_SAVE | LOG_PRINT);
}

sub mylog
{
    my $self = shift;

    my $thing = shift;
    chomp $thing;

    my $which = shift;

    my $MYLOG;
    if ($which & LOG_PRINT) {
        print $thing, "\n";
    }

    if ($which & LOG_SAVE) {
        if ($self->{'MYLOG'}) {
            $MYLOG = $self->{'MYLOG'};
        } else {
            unless ($MYLOG) {
                open ($MYLOG, '>>', $self->{'log_path'} . '/' . 
                    $self->{'log_file'}) or die "Couldn't open logfile!\n";
                my $ofh = select $MYLOG;
                $|=1;
                select $ofh;
                $self->{'MYLOG'} = $MYLOG;

                $self->p("Opened log ($self->{'log_path'}/$self->{'log_file'})");
            }
        }
        flock($MYLOG, LOCK_EX);
        print $MYLOG (scalar gmtime), "\t", $thing, "\n" 
            or die "Couldn't print to MYLOG: $!";
        flock($MYLOG, LOCK_UN);
    }
}

sub GetLogName
{
    my $prefix  = shift || die "Invalid prefix passed for log";

    my $logdate = GetLogDate();
    my $logver  = 0;
    my $logname;

    do {
        $logname = $prefix . $logdate . sprintf("%02d", $logver) . '.log';
        $logver++;
    } until (not -e $logname);

    return $logname;
}

sub GetLogDate
{
    my ($sec,$min,$hour,$mday,$mon,$year,,,) = localtime();

    $mon++;
    $year += 1900;

    if ($min  < 10) { $min  = "0$min"  }
    if ($sec  < 10) { $sec  = "0$sec"  }
    if ($hour < 10) { $hour = "0$hour" }
    if ($mday < 10) { $mday = "0$mday" }
    if ($mon  < 10) { $mon  = "0$mon"  }

    my $time = $year . $mon . $mday;

    return $time;
}


scripts/patch-update_trango.pl
Index: update_trango.pl
===================================================================
RCS file: /cvs/scripts/Admin scripts/Trango_Update/update_trango.pl,v
retrieving revision 1.32
retrieving revision 1.33
diff -u -r1.32 -r1.33
--- update_trango.pl	7 Feb 2007 22:17:52 -0000	1.32
+++ update_trango.pl	7 Feb 2007 23:21:26 -0000	1.33
@@ -1,5 +1,5 @@
 #!/usr/bin/perl
-# $RedRiver: update_trango.pl,v 1.32 2007/02/07 22:17:52 andrew Exp $
+# $RedRiver: update_trango.pl,v 1.33 2007/02/07 23:21:26 andrew Exp $
 ########################################################################
 # update_trango.pl *** Updates trango hosts with a new firmware
 #
@@ -167,9 +167,12 @@
                 $host->{retry}--;
             }
             else {
-                $l->sp("Failed");
+                $l->sp("Failed! - Bye $host->{name}");
+                $l->e("Error updating $firmware_type on $host->{name}" . 
+                    "(try $host->{tries})");
                 $t->bye;
-                next;
+                # don't try any other firmware, don't want to reboot
+                last;
             }
 
         }
@@ -339,6 +342,7 @@
 #use YAML;
 use constant LOG_PRINT => 128;
 use constant LOG_SAVE  => 64;
+use constant LOG_ERR   => 1;
 
 DESTROY {
     my $self = shift;
@@ -378,6 +382,12 @@
     return $self->mylog( $m, LOG_SAVE | LOG_PRINT );
 }
 
+sub e {
+    my $self = shift;
+    my $m    = shift;
+    return $self->mylog( $m, LOG_ERR );
+}
+
 sub mylog {
     my $self = shift;
 
@@ -413,6 +423,11 @@
         print $MYLOG ( scalar gmtime ), "\t", $thing, "\n"
           or die "Couldn't print to MYLOG: $!";
         flock( $MYLOG, LOCK_UN );
+    }
+
+    if ( $which & LOG_ERR ) {
+        # XXX Could tie in here to handle some sort of notifications.
+        print STDERR $thing, "\n";
     }
 }
 


Build.PL
use strict;
use warnings;
use Module::Build;
use File::Spec;
# $RedRiver: Build.PL,v 1.8 2007/02/06 21:19:40 andrew Exp $

my $cfg_file = File::Spec->catfile('t', 'tests.cfg');

my $builder = Module::Build->new(
    module_name         => 'Net::Telnet::Trango',
    license             => 'perl',
    dist_author         => 'Andrew Fresh <andrew@cpan.org>',
    dist_version_from   => 'lib/Net/Telnet/Trango.pm',
    requires => {
        'perl'             => '5.6.1',
        'Net::Telnet'      => '0',
    },
    build_requires => {
        'Test::More' => 0,
    },
    add_to_cleanup      => [ 'Net-Telnet-Trango-*', $cfg_file ],
    create_makefile_pl  => 'traditional',
    create_readme       => 1,
    create_html         => 1,
    dynamic_config      => 1,
    auto_features       => {
        YAML_support =>
        {
            description => "Use YAML.pm to write META.yml files",
            requires => { YAML => ' >= 0.35, != 0.49_01 ' },
        },
        manpage_support =>
        {
            description => "Create Unix man pages",
            requires => { 'Pod::Man' => 0 },
        },
        HTML_support =>
        {
            description => "Create HTML documentation",
            requires => { 'Pod::Html' => 0 },
        },
    },
);

get_config();

$builder->create_build_script();

sub get_config {
    my %cfg;
    print <<EOB;

Net::Telnet::Trango needs some additional information and setup to
perform it\'s full suite of tests. 

These tests MAY change settings on the units it connects to, so please
do not use equipment that is in production.

To skip these tests, hit "return".

EOB

    get_ap_config(\%cfg);

    get_sudb_config(\%cfg) if $cfg{AP} && $cfg{AP_PASSWD};

    get_su_config(\%cfg);

    if (open T,">$cfg_file") {
        while (my ($key, $value) = each %cfg) {
            print T $key, "\t", $value,"\n";
        }
        close T;
    }

}

sub get_ap_config {
    my $cfg = shift;

    print 'To run the AP tests, we need an IP and a password of an AP.', "\n";

    $cfg->{AP}        = $builder->prompt("AP IP: ", $cfg->{AP}) || return;
    $cfg->{AP_PASSWD} = $builder->prompt("AP Password:", $cfg->{AP_PASSWD}) 
      || return;
}

sub get_sudb_config {
    my $cfg = shift;

    print <<EOL;
To run the SUDB tests, we need an suid and mac address of an 
SU that can reliably link to the AP.
EOL

    $cfg->{SU_ID}  = $builder->prompt("SU ID: ", $cfg->{SU_ID})  || return;
    $cfg->{SU_MAC} = $builder->prompt("SU MAC:", $cfg->{SU_MAC}) || return;
}

sub get_su_config {
    my $cfg = shift;

    print 'To run the SU tests, we need an IP and a password of an SU.', "\n";

    $cfg->{SU}        = $builder->prompt("SU IP: ", $cfg->{SU}) || return;
    $cfg->{SU_PASSWD} = $builder->prompt("SU Password:", $cfg->{SU_PASSWD}) 
      || return;
}

